src/ripple/nodestore/impl/DatabaseShardImp.cpp

```diff
@@ -303,62 +303,103 @@ DatabaseShardImp::prepareLedger(std::uint32_t validLedgerSeq)
+
+        if (shardIndex < earliestShardIndex())
+        {
+            return fail(
+                "comes before earliest shard index " +
+                    std::to_string(earliestShardIndex()),
+                shardIndex);
+        }
+
+        // If we are synced to the network, check if the shard index
+        // is greater or equal to the current shard.
+        auto seqCheck = [&](std::uint32_t seq) {
+            // seq will be greater than zero if valid
+            if (seq >= earliestLedgerSeq() &&
+                shardIndex >= seqToShardIndex(seq))
+                return fail("has an invalid index", shardIndex);
+            return true;
+        };
+        if (!seqCheck(app_.getLedgerMaster().getValidLedgerIndex() + 1) ||
+            !seqCheck(app_.getLedgerMaster().getCurrentLedgerIndex()))
+        {
+            return false;
+        }
+
+        if (shards_.find(shardIndex) != shards_.end())
+        {
+            JLOG(j_.debug()) << "shard " << shardIndex
+                             << " is already stored or queued for import";
+            return false;
+        }
+
+        // Any shard earlier than the two most recent shards
+        // is a historical shard
+        if (shardIndex < shardBoundaryIndex(lock))
+            ++historicalShardsToPrepare;
     }
 
-    // If we are synced to the network, check if the shard index
-    // is greater or equal to the current shard.
-    auto seqCheck = [&](std::uint32_t seq) {
-        // seq will be greater than zero if valid
-        if (seq >= earliestLedgerSeq() && shardIndex >= seqToShardIndex(seq))
-            return fail("has an invalid index");
-        return true;
-    };
-    if (!seqCheck(app_.getLedgerMaster().getValidLedgerIndex() + 1) ||
-        !seqCheck(app_.getLedgerMaster().getCurrentLedgerIndex()))
+    auto const numHistShards = numHistoricalShards(lock);
+
+    // Check shard count and available storage space
+    if (numHistShards + historicalShardsToPrepare > maxHistoricalShards_)
+        return fail("maximum number of historical shards reached");
+
+    if (historicalShardsToPrepare)
     {
-        return false;
+        // Check available storage space for historical shards
+        if (!sufficientStorage(
+                historicalShardsToPrepare, PathDesignation::historical, lock))
+            return fail("insufficient storage space available");
     }
 
-    if (shards_.find(shardIndex) != shards_.end())
+    if (auto const recentShardsToPrepare =
+            shardIndexes.size() - historicalShardsToPrepare;
+        recentShardsToPrepare)
     {
-        JLOG(j_.debug()) << "shard " << shardIndex
-                         << " is already stored or queued for import";
-        return false;
+        // Check available storage space for recent shards
+        if (!sufficientStorage(
+                recentShardsToPrepare, PathDesignation::none, lock))
+            return fail("insufficient storage space available");
```

Comment:
Not exercised in unit tests.

Reply:
Added unit tests to exercise additional paths, but this specific error condition is not triggered.